ddtrace/internal/context_manager.py

```diff
@@ -0,0 +1,104 @@
+import abc
+import threading
+from ddtrace.vendor import six
+
+from .logger import get_logger
+from ..context import Context
+
+log = get_logger(__name__)
+
+try:
+    from contextvars import ContextVar
+    _DD_CONTEXTVAR = ContextVar('datadog_contextvar', default=None)
+    CONTEXTVARS_IS_AVAILABLE = True
+except ImportError:
+    CONTEXTVARS_IS_AVAILABLE = False
+
+
+class BaseContextManager(six.with_metaclass(abc.ABCMeta)):
+    def __init__(self, reset=True):
+        if reset:
+            self.reset()
+
+    @abc.abstractmethod
+    def _has_active_context(self):
+        pass
+
+    @abc.abstractmethod
+    def set(self, ctx):
+        pass
+
+    @abc.abstractmethod
+    def get(self):
+        pass
+
+    def reset(self):
+        pass
+
+
+class ThreadLocalContext(BaseContextManager):
+    """
+    ThreadLocalContext can be used as a tracer global reference to create
+    a different ``Context`` for each thread. In synchronous tracer, this
+    is required to prevent multiple threads sharing the same ``Context``
+    in different executions.
+    """
+    def __init__(self, reset=True):
+        # always initialize a new thread-local context holder
+        self.reset()
```

Comment:
we should still call the base `__init__`, just maybe:

```python
def __init__(self):
    super(ThreadLocalContext, self).__init__(reset=True)
```

Reply:
Although, probably still need `__init__(self, reset=True)` here, can just override and force it to be `reset=True` on the call to `super()`.